packages/runtime/container-runtime/src/runtimeLayerCompatState.ts

```diff
@@ -0,0 +1,143 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	checkLayerCompatibility,
+	type ILayerCompatDetails,
+	type ILayerCompatSupportRequirements,
+} from "@fluid-internal/client-utils";
+import type { ICriticalContainerError } from "@fluidframework/container-definitions";
+import { UsageError } from "@fluidframework/telemetry-utils/internal";
+
+import { pkgVersion } from "./packageVersion.js";
+
+/**
+ * The core compatibility details of the Runtime layer that is the same across all layer boundaries.
+ * @internal
+ */
+export const RuntimeCoreCompatDetails = {
+	/**
+	 * The package version of the Runtime layer.
+	 */
+	pkgVersion,
+	/**
+	 * The current generation of the Runtime layer.
+	 */
+	generation: 1,
+};
+
+/**
+ * Runtime's compatibility details that is exposed to the Loader layer.
+ * @internal
+ */
+export const RuntimeCompatDetailsForLoader: ILayerCompatDetails = {
+	...RuntimeCoreCompatDetails,
+	/**
+	 * The features supported by the Runtime layer across the Runtime / Loader boundary.
+	 */
+	supportedFeatures: new Set<string>(),
+};
+
+/**
+ * The requirements that the Loader layer must meet to be compatible with this Runtime.
+ * @internal
+ */
+export const LoaderSupportRequirements: ILayerCompatSupportRequirements = {
+	/**
+	 * Minimum generation that Loader must be at to be compatible with Runtime. Note that 0 is used here so
+	 * that Loader layers before the introduction of the layer compatibility enforcement are compatible.
+	 */
+	minSupportedGeneration: 0,
+	/**
+	 * The features that the Loader must support to be compatible with Runtime.
+	 */
+	requiredFeatures: [],
+};
+
+/**
+ * Runtime's compatibility details that is exposed to the DataStore layer.
+ * @internal
+ */
+export const RuntimeCompatDetailsForDataStore: ILayerCompatDetails = {
+	...RuntimeCoreCompatDetails,
+	/**
+	 * The features supported by the Runtime layer across the Runtime / Loader boundary.
+	 */
+	supportedFeatures: new Set<string>(),
+};
+
+/**
+ * The requirements that the DataStore layer must meet to be compatible with this Runtime.
+ * @internal
+ */
+export const DataStoreSupportRequirements: ILayerCompatSupportRequirements = {
```

Comment:
Consider putting these requirements/details in a dedicated file with literally nothing else. Almost as if it's a JSON file (but with comments and type safety and no need to parse)

Reply:
Do you mean just move the functions to a different file? What's the advantage of doing that?

Reply:
Advantage is to separate data from code

Reply:
Okay. I will consider doing that in one of my follow up PRs. I will merge this one for now.